wrappers/services.go

```diff
@@ -901,33 +902,48 @@ func EnsureSnapServices(snaps map[*snap.Info]*SnapServiceOptions, opts *EnsureSn
 	return context.reloadModified()
 }
 
-// StopServicesFlags carries additional parameters for StopServices.
-// XXX: Rename to StopServicesOptions
-type StopServicesFlags struct {
-	Disable bool
-	ScopeOptions
-}
+// categorizeServices returns a list of system and user services
+// ordered by the same order that 'app' is passed into.
+func categorizeServices(apps []*snap.AppInfo, includeActivatedServices bool) ([]string, []string) {
+	var systemServices []string
+	var userServices []string
 
-// StopServices stops and optionally disables service units for the applications
-// from the snap which are services.
-func StopServices(apps []*snap.AppInfo, flags *StopServicesFlags, reason snap.ServiceStopReason, inter Interacter, tm timings.Measurer) error {
-	sysd := systemd.New(systemd.SystemMode, inter)
-	if flags == nil {
-		flags = &StopServicesFlags{}
+	// process all services of the snap in the order specified by the
+	// caller; before batched calls were introduced, the sockets and timers
+	// were started first, followed by other non-activated services
+	markServices := func(svcs []string, scope snap.DaemonScope) {
+		switch scope {
+		case snap.SystemDaemon:
+			systemServices = append(systemServices, svcs...)
+		case snap.UserDaemon:
+			userServices = append(userServices, svcs...)
+		}
 	}
-
-	if reason != snap.StopReasonOther {
-		logger.Debugf("StopServices called for %q, reason: %v", apps, reason)
-	} else {
-		logger.Debugf("StopServices called for %q", apps)
+	// first, gather all socket and timer units
+	for _, app := range apps {
+		// Get all units for the service, but we only deal with
+		// the activators here.
+		_, activators := internal.SnapServiceUnits(app)
+		if len(activators) == 0 {
+			// just skip if there are no activated units
+			continue
+		}
+		markServices(activators, app.DaemonScope)
 	}
 
-	cli, err := newUserServiceClientNames(flags.Users, inter)
-	if err != nil {
-		return err
+	// now collect all services
+	for _, app := range apps {
+		if serviceIsActivated(app) && !includeActivatedServices {
+			continue
+		}
+		svcName := app.ServiceName()
+		markServices([]string{svcName}, app.DaemonScope)
```

Comment:
and then you can avoid the awkwardness
```suggestion
		collectServices(app.DaemonScope, svcName)
```

Reply:
much nicer, thanks